Content.Shared/Storage/EntitySystems/BinSystem.cs

```diff
@@ -29,150 +30,153 @@ public override void Initialize()
     {
-        if (handled || !canReach)
-            return;
-
-        if (!TryInsertIntoBin(target, itemInHand, component))
-            return;
+        if (entity.Comp.ItemContainer.Count >= entity.Comp.MaxItems)
+            return false;
 
-        _admin.Add(LogType.Pickup, LogImpact.Low, $"{ToPrettyString(target):player} inserted {ToPrettyString(user)} into bin {ToPrettyString(target)}.");
+        return _whitelistSystem.IsWhitelistPassOrNull(entity.Comp.Whitelist, toInsert);
     }
 
     /// <summary>
     /// Inserts an entity at the top of the bin
     /// </summary>
-    /// <param name="uid"></param>
-    /// <param name="toInsert"></param>
-    /// <param name="component"></param>
-    /// <returns></returns>
-    public bool TryInsertIntoBin(EntityUid uid, EntityUid toInsert, BinComponent? component = null)
+    /// <param name="entity">Bin entity we're trying to insert into</param>
+    /// <param name="toInsert">Entity we're trying to insert</param>
+    /// <param name="user">Entity who is inserting into the bin if one exists.</param>
+    /// <returns>Returns true if insertion was successful.</returns>
+    public bool TryInsertIntoBin(Entity<BinComponent> entity, EntityUid toInsert, EntityUid? user = null)
     {
-        if (!Resolve(uid, ref component))
+        if (!CanInsertIntoBin(entity, toInsert))
             return false;
 
-        if (component.Items.Count >= component.MaxItems)
-            return false;
+        _container.Insert(toInsert, entity.Comp.ItemContainer);
 
-        if (_whitelistSystem.IsWhitelistFail(component.Whitelist, toInsert))
-            return false;
+        if (user != null)
+        {
+            _admin.Add(LogType.Pickup,
+                LogImpact.Low,
+                $"{ToPrettyString(user):player} inserted {ToPrettyString(toInsert)} into bin {ToPrettyString(entity)}.");
+        }
 
-        _container.Insert(toInsert, component.ItemContainer);
-        Dirty(uid, component);
         return true;
     }
 
     /// <summary>
     /// Tries to remove an entity from the top of the bin.
     /// </summary>
-    /// <param name="uid"></param>
-    /// <param name="toRemove"></param>
-    /// <param name="component"></param>
-    /// <returns></returns>
-    public bool TryRemoveFromBin(EntityUid uid, EntityUid? toRemove, BinComponent? component = null)
+    /// <param name="entity">Entity we're removing an item from.</param>
+    /// <param name="removed">Item we removed.</param>
+    /// <param name="user">Entity that is trying to remove from the bin.</param>
+    /// <returns>Returns false if removal was successful.</returns>
+    public bool TryRemoveFromBin(Entity<BinComponent> entity, [NotNullWhen(true)] out EntityUid? removed, EntityUid? user = null)
     {
-        if (!Resolve(uid, ref component))
-            return false;
+        removed = entity.Comp.ItemContainer.ContainedEntities.LastOrDefault();
```

Comment:
EntityUid is a value type so its default is a uid of 0; this needs to either have the `.Cast<EntityUid?>()` or a check that the resulting EntityUid is valid. Ditto for the other. Ditto for the other LastOrDefault.